### PR TITLE
curl: add hyper support

### LINKED ISSUE #357409: curl: move to hyper backend
## Project description

Hyper and curl are [looking for Linux distributions](https://seanmonstar.com/blog/hyper-in-curl-needs-a-champion/) to ship curl with hyper as an HTTP backend. The work is almost done, but not quite and they want someone (a Linux distro, e.g.) to commit to shipping it.

We already include rustls as an optional dependency for curl. Supporting hyper-ffi seems fairly similar.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by getchoo**
~~Closes https://github.com/NixOS/nixpkgs/issues/357409~~

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by JohnRTitor — COMMENTED** [review]
since it's actually `false`, we are not shipping curl with hyper, unlike what [the post](https://seanmonstar.com/blog/hyper-in-curl-needs-a-champion/) wants.

I would let the issue stay open and change the issue title to "curl: move to the hyper backend".
I think it's great to add the support by default, once it is stable, that is.

File: pkgs/by-name/li/libhyper/package.nix [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    repo = "hyper";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-neFCZfMWm+FV+VlSXpvrxaGj+wF+exVDZv3JG4wiX7o=";
+  };
+
+  # Add a Cargo.lock
+  #
+  # Make sure rustc always emits cdylib artifacts
+  postPatch = ''
+    ln -sf ${./Cargo.lock} Cargo.lock
```
Comment (JohnRTitor) [inline pkgs/by-name/li/libhyper/package.nix:33]: Upstream does not commit their lockfiles?

I think a better approach would be to ask upstream to commit their lockfile, mentioning our reproducibility requirements which needs the lockfile to compare the hash with.

Comment (pluiedev) [reply pkgs/by-name/li/libhyper/package.nix:33]: Hyper is usually used as a library, and Cargo's commit guidelines generally advise against including a Cargo.lock for libraries

Comment (getchoo) [reply pkgs/by-name/li/libhyper/package.nix:29]: Opened https://github.com/hyperium/hyper/issues/3788

File: pkgs/by-name/li/libhyper/package.nix [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    RUSTFLAGS = toString [
+      "--cfg"
+      "hyper_unstable_ffi"
+    ];
+  };
+
+  passthru = {
+    updateScript = nix-update-script { };
+  };
+
```
Comment (milibopp) [inline pkgs/by-name/li/libhyper/package.nix:70]: You could test curl compiled against hyper here as a `passthru.test`.

Comment (getchoo) [reply pkgs/by-name/li/libhyper/package.nix:73]: Good idea!

**Comment by niklaskorz** [comment]
Curl has officially removed hyper support: https://daniel.haxx.se/blog/2024/12/21/dropping-hyper/

**Pull request closed without merging**